### PR TITLE
Document Metadata Context Concept

### PR DESCRIPTION
**Pull request by JoergSiebahn**

### CODE REVIEW

File: sda-commons-server-dropwizard/docs/metadata-context.md [hunk 1]
```diff
@@ -0,0 +1,189 @@
+# The Metadata Context
+
+
+## Purpose
+
+The metadata context keeps contextual information of a business process across all involved
+microservices.
+The information that is stored in the metadata context is solely defined by the environment.
+An environment may not define any metadata.
```
Comment (eikenilsknopp) [inline sda-commons-server-dropwizard/docs/metadata-context.md:9]: This sounds confusing, doesn’t the second sentence contradict the first?

File: sda-commons-server-dropwizard/docs/metadata-context.md [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+support features a service in the end of a process provides based on the initial user request.
+
+A service must be able to fulfill its purpose without any information from the metadata context.
+A service may use information from the metadata context to use or apply different business related
+or non-functional configuration for a specific business process but must have a default for these
+configurations as well.
+
+The metadata context is technically transferred via HTTP headers and Kafka message headers between
+services in the same environment.
+At runtime, it is hold in a thread local to be available for the current process.
```
Comment (eikenilsknopp) [inline sda-commons-server-dropwizard/docs/metadata-context.md:29]: Spelling, "hold in a thread local"?

File: sda-commons-server-dropwizard/docs/metadata-context.md (hunk 1 above)
Comment (JoergSiebahn) [reply sda-commons-server-dropwizard/docs/metadata-context.md:9]: Hmmm … felt more clear a few minutes ago.
The second sentence should clarify that it is not needed to define what's include in the metadata context. An environment may or may not define fields for the context. But it is only the environment/customer/operations that has the right to define it. No service can require that such configuration exists.
I will think about that but I'm also happy to get a suggestion.

File: sda-commons-server-dropwizard/README.md [hunk 3]
```diff
@@ -90,6 +90,10 @@ public void initialize(Bootstrap<Configuration> bootstrap) {
 The [`MetadataContextBundle`](./src/main/java/org/sdase/commons/server/dropwizard/bundles/MetadataContextBundle.java)
 enables the metadata context handling for an application.
 
+If you want to make use of the data in the metadata context, you should read the [dedicated documentation](./docs/metadata-context.md).
```
Comment (christopher-cudennec) [inline sda-commons-server-dropwizard/README.md:93]: I like that it's a dedicated document! 👍

File: sda-commons-server-dropwizard/docs/metadata-context.md (hunk 1 above)
Comment (christopher-cudennec) [reply sda-commons-server-dropwizard/docs/metadata-context.md:9]: I think "may not" would be translated to "darf nicht". After reading your explanation we could rephrase to "it's optional" or "it's not mandatory".

File: sda-commons-server-dropwizard/docs/metadata-context.md [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+microservices.
+The information that is stored in the metadata context is solely defined by the environment.
+An environment may not define any metadata.
+In this case, no metadata will be available.
+
+The metadata should not affect the core business logic of a service but can be used to affect how a
+service communicates with the user (e.g. communication methods, templates or domains used for links)
+or how statistical data is evaluated.
+As the metadata is tracked for a whole business process across all involved services, it supports
+that services which are not in a synchronous call chain, can behave different based on the entry
```
Comment (christopher-cudennec) [inline sda-commons-server-dropwizard/docs/metadata-context.md:16]: ... can behave different**ly** ...

File: sda-commons-server-dropwizard/docs/metadata-context.md (hunk 2 above)
Comment (christopher-cudennec) [reply sda-commons-server-dropwizard/docs/metadata-context.md:29]: past participle form would be "held"

File: sda-commons-server-dropwizard/docs/metadata-context.md [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+No standard component should create a metadata context initially, because the metadata context
+always depends on the environment.
+
+The easiest way to initially create a metadata context is to add the respective headers in a proxy
+that exposes a service, e.g. an Ingress in Kubernetes.
+The proxy may derive the metadata information from the used domain, query params, user agent
+headers and other information available in the request.
+
+If getting the metadata information is more complex, an environment specific interaction service
+(according the SDA service architecture) may be implemented and create the context programmatically.
```
Comment (christopher-cudennec) [inline sda-commons-server-dropwizard/docs/metadata-context.md:63]: ... according **to** ...

File: sda-commons-server-dropwizard/docs/metadata-context.md [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+This service should act like proxy and forward the request to a standard business service.
```
Comment (christopher-cudennec) [inline sda-commons-server-dropwizard/docs/metadata-context.md:64]: ... act like **a** proxy ...

File: sda-commons-server-dropwizard/docs/metadata-context.md [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+Every configuration that can be modified by information of the metadata context must have a default
+as well, because the service may be used in an environment without configured metadata context or in
+a specific business process is no context available. 
+How a service behaves and is configured when metadata context is available must be described in the
+deployment documentation to make this complex flexibility manageable by the operations team.
+
+
+## Example: Template selection
+
+Image a business case that is advertised on different landing pages, where the user enters
```
Comment (christopher-cudennec) [inline sda-commons-server-dropwizard/docs/metadata-context.md:115]: Imagine ...

**Review by christopher-cudennec — COMMENTED** [review]
The most valuable part of the documentation is the concrete example. Maybe we can start with it and explain technical details after that? I'm afraid that the rest might be to abstract otherwise.

**Review by christopher-cudennec — APPROVED** [review]

**Pull request merged**